internal/binder/function/funcs_array.go

```diff
@@ -420,4 +427,174 @@ func registerArrayFunc() {
+			array, ok := args[0].([]interface{})
+			if !ok {
+				return errorArrayFirstArgumentNotArrayError, false
+			}
+
+			var output []interface{}
+
+			for _, val := range array {
+				innerArr, ok := val.([]interface{})
+				if !ok {
+					return errorArrayNotArrayElementError, false
+				}
+				output = append(output, innerArr...)
+			}
+
+			return output, true
+		},
+		val: func(ctx api.FunctionContext, args []ast.Expr) error {
+			return ValidateLen(1, len(args))
+		},
+	}
+	builtins["array_distinct"] = builtinFunc{
+		fType: ast.FuncTypeScalar,
+		exec: func(ctx api.FunctionContext, args []interface{}) (interface{}, bool) {
+			array, ok := args[0].([]interface{})
+			if !ok {
+				return errorArrayFirstArgumentNotArrayError, false
+			}
+
+			output := make([]interface{}, 0, len(array))
+			set := make(map[interface{}]bool)
+
+			for _, val := range array {
+				if !set[val] {
+					output = append(output, val)
+					set[val] = true
+				}
+			}
+
+			return output, true
+		},
+		val: func(ctx api.FunctionContext, args []ast.Expr) error {
+			return ValidateLen(1, len(args))
+		},
+	}
+	builtins["array_map"] = builtinFunc{
+		fType: ast.FuncTypeScalar,
+		exec: func(ctx api.FunctionContext, args []interface{}) (interface{}, bool) {
+			funcName, ok := args[0].(string)
+			if !ok {
+				return errorArrayFirstArgumentNotStringError, false
+			}
+
+			array, ok := args[1].([]interface{})
+			if !ok {
+				return errorArraySecondArgumentNotArrayError, false
+			}
+
+			mapped := make([]interface{}, 0, len(array))
+			var result interface{}
+			for _, v := range array {
+				params := []interface{}{v}
+				fs, ok := builtins[funcName]
+				if !ok {
+					return fmt.Errorf("unknown built-in function: %s.", funcName), false
+				}
+
+				if fs.fType != ast.FuncTypeScalar {
+					return fmt.Errorf("first argument should be a scalar function."), false
+				}
+				eargs := make([]ast.Expr, len(params))
+				if err := fs.val(nil, eargs); err != nil {
+					return fmt.Errorf("function should accept exactly one argument."), false
+				}
+
+				result, ok = fs.exec(ctx, params)
+				if !ok {
+					return result, false
+				}
+				mapped = append(mapped, result)
+			}
+
+			return mapped, true
+		},
+		val: func(ctx api.FunctionContext, args []ast.Expr) error {
+			return ValidateLen(2, len(args))
+		},
+	}
+	builtins["array_join"] = builtinFunc{
+		fType: ast.FuncTypeScalar,
+		exec: func(ctx api.FunctionContext, args []interface{}) (interface{}, bool) {
+			array, ok := args[0].([]interface{})
+			if !ok {
+				return errorArrayFirstArgumentNotArrayError, false
+			}
+
+			delimiter, ok := args[1].(string)
+			if !ok {
+				return errorArraySecondArgumentNotStringError, false
+			}
+
+			var nullReplacement string
+			if len(args) == 3 {
+				nullReplacement, ok = args[2].(string)
+				if !ok {
+					return errorArrayThirdArgumentNotStringError, false
+				}
+			}
+
+			for i, v := range array {
```

Comment:
This will give the unexpected output as `'a::b'`for `array_join(['a',null,'b'],':')`

Reply:
Do you mean that only if nullReplacement exists, the null value in the array will be replaced? For example, the result of `array_join(['a',null,'b'],':')` should be: `'a:b'`?

Reply:
I think so.